Library/src/Multiplayer/EventBus.cpp

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2023 Magnopus LLC
+
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "CSP/Multiplayer/EventBus.h"
+
+#include "CSP/Systems/SystemBase.h"
+#include "CallHelpers.h"
+#include "Debug/Logging.h"
+#include "Memory/Memory.h"
+#include "Multiplayer/SignalR/SignalRConnection.h"
+#include "NetworkEventManagerImpl.h"
+
+namespace csp::multiplayer
+{
+
+extern ErrorCode ParseError(std::exception_ptr Exception);
+
+constexpr const uint64_t ALL_CLIENTS_ID = -1;
+
+EventBus::~EventBus()
+{
+	CSP_DELETE(MultiplayerConnectionInst);
+}
+
+EventBus::EventBus(const EventBus& InBoundConnection)
+{
+	MultiplayerConnectionInst = InBoundConnection.MultiplayerConnectionInst;
+	EventMap				  = InBoundConnection.EventMap;
+}
+
+EventBus::EventBus(MultiplayerConnection* InMultiplayerConnection)
+{
+	MultiplayerConnectionInst = InMultiplayerConnection;
+	EventMap				  = {};
+}
+
+void EventBus::ListenEvent(const csp::common::String& EventName, ParameterisedCallbackHandler Callback, csp::systems::SystemBase* System)
+{
+	if (MultiplayerConnectionInst->Connection == nullptr || !MultiplayerConnectionInst->Connected)
+	{
+		return;
+	}
+
+	if (!Callback && !System)
+		return;
+
+	if (Callback)
+	{
+		EventMap[EventName].first.push_back(Callback);
+	}
+	if (System)
+	{
+		EventMap[EventName].second = System;
+	}
+}
+
+void EventBus::ListenEvent(const csp::common::String& EventName, csp::systems::SystemBase* System)
+{
+	ListenEvent(EventName, nullptr, System);
+}
+
+void EventBus::ListenEvent(const csp::common::String& EventName, ParameterisedCallbackHandler Callback)
+{
+	ListenEvent(EventName, Callback, nullptr);
+}
+
+void EventBus::StopListenEvent(const csp::common::String& EventName)
+{
+	if (!EventMap.empty() && EventMap.find(EventName) != EventMap.end())
```

Comment:
The issue here is that it doesn't care if multiple callbacks or a system have been registered. So if multiple callbacks have been registered and a registered system calls this, it will mean the registered callbacks are also removed.

Reply:
that is a very good point, I did not realise this would be a use case. what is the best solution, passing the callback or system?

Reply:
It would also be an issue if just multiple callbacks were registered and no system. If this method were called for one callback then all callbacks would be removed.

I think this could be a valid method - to stope everything listening for an event, but would likely need to rename/make it very clear what it did.

If we are supporting registering multiple callbacks (or a callback/system combo) with a single event, then we will need a method for unregistering them individually.